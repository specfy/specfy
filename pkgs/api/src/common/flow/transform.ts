import type { Edge, Node } from 'reactflow';

import type {
  ComponentForFlow,
  ComputedFlow,
  EdgeData,
  NodeData,
} from './types.js';

export function createNode(component: ComponentForFlow): Node<NodeData> {
  const node: Node<NodeData> = {
    id: component.id,
    type: 'custom',
    data: {
      label: component.name,
      type: component.type,
      techId: component.techId,
      originalSize: component.display.size,
    },
    position: { ...component.display.pos },
    targetPosition: 'left' as any,
    sourcePosition: 'right' as any,
    style: {
      width: `${component.display.size.width}px`,
      height: `${component.display.size.height}px`,
    },
    deletable: true,
    connectable: true,
    draggable: true,
    focusable: true,
    selectable: true,
  };

  if (component.inComponent) {
    node.extent = 'parent';
    node.parentNode = component.inComponent;
  }

  return node;
}

export function getEdgeMarkers(data: EdgeData) {
  const edge: Partial<Edge> = {};
  if (data.read) {
    edge.markerStart = {
      type: 'arrowclosed' as any,
      width: 10,
      height: 10,
    };
  }
  if (data.write) {
    edge.markerEnd = {
      type: 'arrowclosed' as any,
      width: 10,
      height: 10,
    };
  }
  return edge;
}
export function componentsToFlow(components: ComponentForFlow[]): ComputedFlow {
  const edges: Array<Edge<EdgeData>> = [];
  const nodes: Array<Node<NodeData>> = [];

  // Create all hosting nodes
  // We need to add them first because React Flow is not reordering
  const hosts = components.filter((comp) => comp.type === 'hosting');
  const done: string[] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    if (hosts.length <= 0) {
      break;
    }
    const host = hosts.shift()!;
    if (host.inComponent && !done.includes(host.inComponent)) {
      hosts.push(host);
      continue;
    }

    nodes.push(createNode(host));
    done.push(host.id);
  }

  // Create all other nodes
  for (const comp of components) {
    if (comp.type === 'hosting') {
      continue;
    }

    nodes.push(createNode(comp));
  }

  for (const comp of components) {
    for (const edge of comp.edges) {
      const item: Edge<EdgeData> = {
        id: `${comp.id}->${edge.to}`,
        source: comp.id,
        target: edge.to,
        sourceHandle: edge.portSource,
        targetHandle: edge.portTarget,
        data: { read: edge.read, write: edge.write },
        ...getEdgeMarkers(edge),
        // type: 'floating',
      };
      edges.push(item);
    }
  }

  return { edges, nodes };
}
