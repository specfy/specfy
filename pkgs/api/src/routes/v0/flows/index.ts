import type { FastifyPluginCallback } from 'fastify';

import get from './get.js';

const fn: FastifyPluginCallback = async (f, _, done) => {
  f.register(get, { prefix: '/flows' });

  done();
};

export default fn;
