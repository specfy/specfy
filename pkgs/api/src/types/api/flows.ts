import type { ComputedFlow } from '../../common/flow/types.js';

import type { QuerystringOrgMaybeProject, Res } from './api.js';

export interface ApiFlow {
  id: string;

  flow: ComputedFlow;

  createdAt: string;
  updatedAt: string;
}

// GET /:id
export type GetFlow = Res<{
  Querystring: QuerystringOrgMaybeProject;
  Success: {
    data: ApiFlow;
  };
}>;
