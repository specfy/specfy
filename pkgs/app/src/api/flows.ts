import type { GetFlow, PatchFlow } from '@specfy/models';
import { useQuery } from '@tanstack/react-query';

import { qcli } from '../common/query';

import { fetchApi } from './fetch';
import { APIError, isError } from './helpers';

export function useGetFlow({ flow_id, org_id, project_id }: GetFlow['QP']) {
  return useQuery({
    queryKey: ['getFlow', org_id, project_id, flow_id],
    queryFn: async (): Promise<GetFlow['Success']> => {
      const { json, res } = await fetchApi<GetFlow>(`/flows/${flow_id}`, {
        qp: { org_id, project_id },
      });

      if (res.status !== 200 || isError(json)) {
        throw new APIError({ res, json });
      }

      return json;
    },
  });
}

export async function updateFlow(
  { flow_id, org_id, project_id }: PatchFlow['QP'],
  data: PatchFlow['Body']
): Promise<PatchFlow['Reply']> {
  const { res, json } = await fetchApi<PatchFlow>(
    `/flows/${flow_id}`,
    { body: data, qp: { org_id, project_id } },
    'PATCH'
  );

  if (res.status === 200) {
    void qcli.invalidateQueries(['getFlow', org_id]);
  }

  return json;
}
