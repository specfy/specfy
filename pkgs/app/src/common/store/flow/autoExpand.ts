import type { ComputedNode } from '@specfy/models';

/**
 * Automatically expand host nodes when dragging a node inside.
 * Right now it's only pushing boundaries not the node outside
 */
export function autoExpand(node: ComputedNode, nodes: ComputedNode[]): void {
  const pad = 5;
  const par = nodes.find((n) => n.id === node.parentNode)!;
  const pos = par.position;

  const exWidth = node.position.x + node.width! - (par.width! - pad);
  const exHeight = node.position.y + node.height! - (par.height! - pad);

  if (exWidth > 0) {
    // Pushing RIGHT
    par.width! += exWidth;
  } else if (node.position.x < pad) {
    // Pushing LEFT
    const incr = Math.abs(node.position.x);
    par.width! += incr;
    par.position.x = pos.x - incr;
    node.position.x = pad;
  }

  if (exHeight > 0) {
    // Pushing BOT
    par.height! += exHeight;
  } else if (node.position.y < pad) {
    // Pushing TOP
    const incr = Math.abs(node.position.y);
    par.height! += incr;
    par.position.y = par.position.y - incr;
    node.position.y = pad;
  }

  par.width = Math.round(par.width!);
  par.height = Math.round(par.height!);
  par.style!.width = par.width!;
  par.style!.height = par.height!;
}
