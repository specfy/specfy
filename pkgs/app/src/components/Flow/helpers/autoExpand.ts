import type { ComputedFlow, ComputedNode } from '@specfy/models';

import type { BatchNodeUpdate } from '../types';

/**
 * Automatically expand host nodes when dragging a node inside.
 * Right now it's only pushing boundaries not the node outside
 */
export function autoExpand(
  flow: ComputedFlow,
  node: ComputedNode,
  evt: Pick<React.MouseEvent<Element, MouseEvent>, 'movementX' | 'movementY'>
) {
  const pad = 20;
  const par = flow.nodes.find((n) => n.id === node.parentNode)!;
  const changes: BatchNodeUpdate[] = [];
  const size = par.data.originalSize;
  const pos = par.position;
  let width: number | undefined;
  let height: number | undefined;
  let x: number | undefined;
  let y: number | undefined;

  if (evt.movementX > 0) {
    // Pushing RIGHT
    const xR = node.position.x + node.data.originalSize.width;
    const diff = xR >= size.width! - pad;

    if (diff) {
      width = size.width + Math.max(xR - (size.width - pad), evt.movementX);
    }
  } else if (evt.movementX < 0) {
    // Pushing LEFT
    const diff = node.position.x < pad;
    const incr = Math.round(
      Math.max(5, Math.abs(evt.movementX), pad - node.position.x)
    );

    if (diff) {
      width = Math.max(size.width + incr, size.width);
      x = pos.x - incr;
    }
  }
  if (evt.movementY > 0) {
    // Pushing BOT
    const yB = node.position.y + node.data.originalSize.height;
    const diff = yB > size.height - pad;

    if (diff) {
      height = Math.max(
        size.height + Math.max(yB - (size.height - pad), evt.movementY),
        size.height
      );
    }
  } else if (evt.movementY < 0) {
    // Pushing TOP
    const diff = node.position.y < pad;
    const incr = Math.round(
      Math.max(5, Math.abs(evt.movementY), pad - node.position.y)
    );

    if (diff) {
      height = Math.max(size.height + incr, size.height);
      y = pos.y - incr;
    }
  }

  if (width || height || x || y) {
    // we first modify the host
    changes.push({
      id: par.id,
      size: {
        width: width ? Math.round(width) : size.width,
        height: height ? Math.round(height) : size.height,
      },
      position: { x: x || pos.x, y: y || pos.y },
    });

    if (x || y) {
      // Then try to reposition the node a the correct spot
      changes.push({
        id: node.id,
        size: { ...node.data.originalSize },
        position: {
          y: y ? pad + 1 : node.position.y,
          x: x ? pad + 1 : node.position.x,
        },
      });
    }
  }

  return changes;
}
