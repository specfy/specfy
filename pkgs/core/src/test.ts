// Nothing in this file work
// I don't know why

// import { config } from 'dotenv';
// import { expect } from 'vitest';

// config({
//   path: './.env',
// });

console.log('top');

// const dateReg = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z$/;

// expect.extend({
//   toBeIsoDate: (received: any) => {
//     if (!dateReg.test(received)) {
//       return {
//         message: () => `expected ${received} to be an ISO Date`,
//         pass: false,
//       };
//     }
//     return { pass: true, message: () => '' };
//   },
// });
