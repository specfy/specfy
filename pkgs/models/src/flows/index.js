export * from './constants.js';
export * from './helpers.js';
export * from './types.api.js';
export * from './types.js';
