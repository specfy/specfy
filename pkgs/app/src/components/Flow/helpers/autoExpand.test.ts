import { getBlobComponent } from '@specfy/models/src/components/test.utils';
import { createNode } from '@specfy/models/src/flows/transform';
import { describe, expect, it } from 'vitest';

import { autoExpand } from './autoExpand';

describe('autoExpand', () => {
  it('should return empty updates', () => {
    const host = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    const node = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    node.parentNode = host.id;
    const updates = autoExpand({ nodes: [host, node], edges: [] }, node, {
      movementX: 0,
      movementY: 0,
    });
    expect(updates).toStrictEqual([]);
  });

  it('should push left', () => {
    const host = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    host.id = 'a';
    const node = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    node.id = 'b';
    node.parentNode = host.id;
    const updates = autoExpand({ nodes: [host, node], edges: [] }, node, {
      movementX: -1,
      movementY: 0,
    });
    expect(updates).toMatchSnapshot();
  });

  it('should push right', () => {
    const host = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    host.id = 'a';
    const node = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    node.id = 'b';
    node.parentNode = host.id;
    const updates = autoExpand({ nodes: [host, node], edges: [] }, node, {
      movementX: 100,
      movementY: 0,
    });
    expect(updates).toMatchSnapshot();
  });

  it('should push top', () => {
    const host = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    host.id = 'a';
    const node = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    node.id = 'b';
    node.parentNode = host.id;
    const updates = autoExpand({ nodes: [host, node], edges: [] }, node, {
      movementX: 0,
      movementY: -1,
    });
    expect(updates).toMatchSnapshot();
  });

  it('should push bot', () => {
    const host = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    host.id = 'a';
    const node = createNode(getBlobComponent({ id: 'project', orgId: 'acme' }));
    node.id = 'b';
    node.parentNode = host.id;
    const updates = autoExpand({ nodes: [host, node], edges: [] }, node, {
      movementX: 0,
      movementY: 100,
    });
    expect(updates).toMatchSnapshot();
  });
});
