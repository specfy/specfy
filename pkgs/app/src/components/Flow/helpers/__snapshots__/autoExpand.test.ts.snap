// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autoExpand > should push bot 1`] = `
[
  {
    "id": "a",
    "position": {
      "x": 0,
      "y": 0,
    },
    "size": {
      "height": 140,
      "width": 130,
    },
  },
]
`;

exports[`autoExpand > should push left 1`] = `
[
  {
    "id": "a",
    "position": {
      "x": -20,
      "y": 0,
    },
    "size": {
      "height": 40,
      "width": 150,
    },
  },
  {
    "id": "b",
    "position": {
      "x": 21,
      "y": 0,
    },
    "size": {
      "height": 40,
      "width": 130,
    },
  },
]
`;

exports[`autoExpand > should push right 1`] = `
[
  {
    "id": "a",
    "position": {
      "x": 0,
      "y": 0,
    },
    "size": {
      "height": 40,
      "width": 230,
    },
  },
]
`;

exports[`autoExpand > should push top 1`] = `
[
  {
    "id": "a",
    "position": {
      "x": 0,
      "y": -20,
    },
    "size": {
      "height": 60,
      "width": 130,
    },
  },
  {
    "id": "b",
    "position": {
      "x": 0,
      "y": 21,
    },
    "size": {
      "height": 40,
      "width": 130,
    },
  },
]
`;
